Introduce concrete error type

jwass opened the pull request:
Addresses #417. Still a work in progress.
Remaining questions:
- [x] * Documentation on new error type
- [x] * Should we remove `pub use rayon_core::initialize` in `rayon/src/lib.rs`
- [x] * Finalize name decision for `ThreadPoolInitError` and `Configuration::init()`

Define ThreadPoolInitError. Deprecate methods that returned
Box<Error> types in favor of new methods returning the concrete
ThreadPoolInitError.

Deprecations in rayon-core:
initialize() => Configuration::init_global()
Configuration::build() => Configuration::init()
ThreadPool::new() => Configuration::init()

cuviper commented on rayon-core/src/join/test.rs:
```diff
@@ -44,7 +43,7 @@ fn sort_in_pool() {
     let mut data: Vec<_> = (0..12 * 1024).map(|_| rng.next_u32()).collect();
 
-    let pool = ThreadPool::new(Configuration::new()).unwrap();
+    let pool = Configuration::new().init().unwrap();
```
(bike-shedding) It doesn't seem great that this call says nothing about being a thread pool -- though this was true of `::build()` too.  Maybe it's OK for folks who would call it like `rayon::Configuration::new().init()` so the context isn't entirely lost.

cuviper commented on rayon-core/src/lib.rs:
```diff
@@ -84,6 +85,18 @@ pub fn current_num_threads() -> usize {
+#[derive(Debug)]
+pub struct ThreadPoolInitError {
+    kind: ErrorKind,
+
```
rogue blank line

cuviper commented on rayon-core/src/lib.rs:
```diff
@@ -306,27 +347,30 @@ impl Configuration {
+
+impl fmt::Display for ThreadPoolInitError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        self.description().fmt(f)
```
`io::Error` is more informative in `Display`, so it would be nice to forward that directly.

jwass replied:
Like this? https://github.com/rayon-rs/rayon/pull/516/files#diff-f42c354e8416b0ad30f65243a93d84a0R369

jwass commented on rayon-core/src/test.rs:
```diff
@@ -118,3 +117,19 @@ fn handler_panics_handled_correctly() {
+}
+
+
+/// Helper used by check_error_send_sync to ensure ThreadPoolInitError is Send + Sync
```
Is this the right way to write a test to ensure the type is Send + Sync?

jwass commented:
@cuviper Addressed all the requested changes + a few other things - with the exception of coming up with a final name for `Configuration::init`. I figure we can decide on that once the rest of the code is good to go.

cuviper commented:
Another option we were discussing on gitter is to rename `Configuration` to something like `ThreadPoolBuilder`.  (Or really a "soft" rename that still leaves a deprecated `Configuration` behind.)  Then I think we'll have better context for what a `build()`/`init()` method is doing.

jwass commented:
@cuviper @nikomatsakis 
Here's a first pass with the new `ThreadPoolBuilder` with `Configuration` delegating to it.

* Rename `Configuration` to `ThreadPoolBuilder` and add `build(self) -> Result<ThreadPool, ThreadPoolInitError>`
* Make a deprecated `Configuration` that's a thin wrapper around `ThreadPoolBuilder` while maintaining its original interface.

Still left to do:
* [ ] Update any lingering `Configuration` references in the documentation.
* [ ] Add some tests to ensure the deprecated `Configuration` is still fully supported.

If you agree with the approach I'll finish out the rest.

cuviper commented:
> Add some tests to ensure the deprecated Configuration is still fully supported.

I've had a vague idea that we ought to be testing past eras of `rayon`.  We could create some more local repos under `ci/` that require old `rayon` versions, but override their `rayon-core` to the local path, then make sure they at least still build.  (We don't need to make this idea part of your PR though.)

jwass commented on rayon-core/src/lib.rs:
```diff
@@ -128,15 +148,38 @@ type StartHandler = Fn(usize) + Send + Sync;
+    /// changed. Therefore, if you call `init_global` a second time, it
+    /// will return an error. An `Ok` result indicates that this
+    /// is the first initialization of the thread pool.
+    pub fn init_global(self) -> Result<(), ThreadPoolInitError> {
```
That was exactly my thinking to leaving it as `init_global` - but I have no objections to `build_global` or anything else if that's what makes sense. I'll defer to your judgment.

nikomatsakis commented on README.md:
```diff
@@ -93,7 +93,7 @@ equals to the number of logical cores and not the physical ones.
 If you want to alter the number of threads spawned, you can set the
 environmental variable `RAYON_NUM_THREADS` to the desired number of threads
-or use the [`initialize` function](https://docs.rs/rayon/*/rayon/fn.initialize.html) method
+or use the [`Configuration::init_global` function](https://docs.rs/rayon/*/rayon/struct.Configuration.html#method.init_global) method
```
Nit: it's not `ThreadPoolBuilder::init_global`, right?

nikomatsakis commented on rayon-core/src/lib.rs:
```diff
@@ -84,9 +85,21 @@ pub fn current_num_threads() -> usize {
+    IOError(io::Error),
+}
+
 /// Contains the rayon thread pool configuration.
```
I think we could do a bit better here in this documentation. =) Maybe something like this:

---

Used to configure and create a new thread-pool, or to configure the global rayon thread-pool. 

# Examples

The following creates a Rayon thread-pool using 22 threads:

```rust
let pool = ThreadPoolBuilder::new().num_threads(22).build();
```

You can also configure the global thread pool by using `init_global` instead of `build`:

```rust
ThreadPoolBuilder::new().num_threads(22).init_global();
```

jwass commented:
@nikomatsakis I think I addressed everything:
* Renamed `init_global` to `build_global`
* Updated all `Configuration` references, both in code and docs, to use `ThreadPoolBuilder`. (At least I hope I got them all.)
* Added a small test to ensure that using all public methods on `Configuration` at least compile, now that it's deprecated. We can do something more robust like @cuviper's suggestion above but maybe a different PR.
* Fleshed out docs for `ThreadPoolBuilder`.
* Updated `rayon-futures` to use `ThreadPoolBuilder`.
* No compiler warnings.

I'd like to rebase/squash before merging.

cuviper commented:
Looks great to me!

jwass commented:
Rebased/squashed and ready to go.

cuviper commented:
bors r+
